controllers/harborconfiguration_controller.go

```diff
@@ -0,0 +1,421 @@
+		return ctrl.Result{}, err
+	}
+
+	client, err := apiv2.NewRESTClientForHost(getHarborURL(&harborTarget), harborConfiguration.Spec.HarborTarget.HarborUsername, haborSecret, nil)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if harborConfiguration.ObjectMeta.DeletionTimestamp.IsZero() {
+		_, err = r.reconcileAll(ctx, harborConfiguration, client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		_, err = triggerReplication(ctx, harborConfiguration, client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		_, err = deleteAll(ctx, harborConfiguration, client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HarborConfigurationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&harborconfigurationv1alpha1.HarborConfiguration{}).
+		Complete(r)
+}
+
+func (r *HarborConfigurationReconciler) registryReconciliation(ctx context.Context, harborConfiguration harborconfigurationv1alpha1.HarborConfiguration, registry modelv2.Registry, client *apiv2.RESTClient) (ctrl.Result, error) {
+	srcRegistry, err := client.GetRegistryByName(ctx, harborConfiguration.Spec.Replication.RegistryName)
+	hErr := &harborerrors.ErrRegistryNotFound{}
+
+	if err != nil && errors.Is(err, hErr) {
+		err = client.NewRegistry(ctx, &registry)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else if err == nil {
+		update := &modelv2.RegistryUpdate{
+			Name:        &harborConfiguration.Spec.Registry.Name,
+			URL:         &harborConfiguration.Spec.Registry.TargetRegistryUrl,
+			Description: &harborConfiguration.Spec.Registry.Description,
+		}
+		if harborConfiguration.Spec.Registry.Credential != nil {
+			update.AccessKey = &harborConfiguration.Spec.Registry.Credential.AccessKey
+			update.AccessSecret = &harborConfiguration.Spec.Registry.Credential.AccessSecret
+			update.CredentialType = &harborConfiguration.Spec.Registry.Credential.Type
+		}
+		err = client.UpdateRegistry(ctx, update, srcRegistry.ID)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *HarborConfigurationReconciler) projectReconciliation(ctx context.Context, harborConfiguration harborconfigurationv1alpha1.HarborConfiguration, registry modelv2.Registry, client *apiv2.RESTClient) (ctrl.Result, error) {
+	srcRegistry, err := client.GetRegistryByName(ctx, harborConfiguration.Spec.Replication.RegistryName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	project := &modelv2.ProjectReq{
+		ProjectName:  harborConfiguration.Spec.ProjectReq.ProjectName,
+		Public:       harborConfiguration.Spec.ProjectReq.IsPublic,
+		StorageLimit: harborConfiguration.Spec.ProjectReq.StorageLimit,
+		RegistryID:   &srcRegistry.ID,
+	}
+
+	_, err = client.GetProject(ctx, project.ProjectName)
+	hErr := &harborerrors.ErrProjectNotFound{}
```

Comment:
You use `&harborerrors.ErrProjectNotFound{}` a few times. It would be more readable to set it as `var harborProjectNotFoundError` on top of the file and reuse it.